portal-test/src/com/liferay/portal/kernel/test/rule/DataGuardTestRuleUtil.java

```diff
@@ -489,6 +509,111 @@ private static Closeable _installTransactionExecutor(
 		};
 	}
 
+	private static boolean _isOrphanResourcePermission(
+			Map<String, PersistedModelLocalService> persistedModelLocalServices,
+			ResourcePermission resourcePermission)
+		throws PortalException {
+
+		if (resourcePermission.getPrimKeyId() == 0) {
+			return false;
+		}
+
+		PersistedModelLocalService persistedModelLocalService =
+			persistedModelLocalServices.get(resourcePermission.getName());
+
+		if (persistedModelLocalService == null) {
+			return false;
+		}
+
+		try {
+			persistedModelLocalService.getPersistedModel(
+				resourcePermission.getPrimKeyId());
+
+			return false;
+		}
+		catch (NoSuchModelException noSuchModelException) {
+		}
+
+		return true;
+	}
+
+	private static void _orphanResourcePermissionDetection(
```

Comment:
@tinatian @jorgediaz-lr I wonder if we really need a new method to do the detection.

The `_autoDeleteAndAssert` method calls `_autoDeleteLeftOvers` first (if `autoDelete == true`), and then re-captures the snapshot and does the assertion.

So, if we just skip `ResourcePermission` in `_autoDeleteLeftOvers`, but don't skip it in the assertion in `_autoDeleteAndAssert`, the left over `ResourcePermission` can still be detected.

Please see https://github.com/dantewang/liferay-portal/pull/2227/commits/f5d8890f7c021b280311621c052b31e61c1fab1d for what I mean.

I think this way does the same thing that Tina suggests in https://github.com/liferay-core-infra/liferay-portal/pull/596#issuecomment-1098507497

Reply:
@jorgediaz-lr Could you check if there's any edge case that the above approach can't cover?